modules/component/src/core/cd-aware/cd-aware_creator.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  combineLatest,
+  EMPTY,
+  NEVER,
+  NextObserver,
+  Observable,
+  PartialObserver,
+  Subject,
+  Subscribable,
+  Subscription,
+} from 'rxjs';
+import {
+  catchError,
+  distinctUntilChanged,
+  map,
+  startWith,
+  switchMap,
+  tap,
+} from 'rxjs/operators';
+import {
+  CdStrategy,
+  DEFAULT_STRATEGY_NAME,
+  StrategySelection,
+} from './strategy';
+
+export interface CdAware<U> extends Subscribable<U> {
+  nextValue: (value: any) => void;
+  nextConfig: (config: string) => void;
+}
+
+/**
+ * class CdAware
+ *
+ * @description
+ * This abstract class holds all the shared logic for the push pipe and the let directive
+ * responsible for change detection
+ * If you extend this class you need to implement how the update of the rendered value happens.
+ * Also custom behaviour is something you need to implement in the extending class
+ */
+export function createCdAware<U>(cfg: {
+  strategies: StrategySelection<U>;
+  resetContextObserver: NextObserver<any>;
+  updateViewContextObserver: PartialObserver<U> & NextObserver<U>;
+}): CdAware<U | undefined | null> {
+  const configSubject = new Subject<string>();
+  const config$: Observable<CdStrategy<U>> = configSubject.pipe(
+    distinctUntilChanged(),
+    startWith(DEFAULT_STRATEGY_NAME),
+    map(
+      (strategy: string): CdStrategy<U> =>
+        cfg.strategies[strategy]
+          ? cfg.strategies[strategy]
+          : cfg.strategies.idle
+    )
+  );
+
+  const observablesSubject = new Subject<Observable<U>>();
+  const observables$$ = observablesSubject.pipe(distinctUntilChanged());
+
+  let prevObservable: Observable<U>;
+  const renderSideEffect$ = combineLatest([observables$$, config$]).pipe(
+    switchMap(([observable$, strategy]) => {
+      if (prevObservable === observable$) {
+        return NEVER;
+      }
+
+      if (observable$ == null) {
+        cfg.updateViewContextObserver.next(observable$);
+        strategy.render();
+        return NEVER;
```

Comment:
Favor `EMPTY` instead. The `swtichMap` cannot complete if the outer observable is still active, so the net result is the same, with the exception that it will clean up resources earlier with `EMPTY`.

Reply:
True! Very good!
Will insert.

Reply:
Done.